qiita_db/util.py

```diff
@@ -2311,3 +2317,337 @@ def send_email(to, subject, body):
+        _df = df[(df.cName == cname)].copy()
+    else:
+        _df = df[(df.cName == cname) & (df.sName == sname)].copy()
+
+    fig, axs = plt.subplots(ncols=2, figsize=(10, 4), sharey=False)
+    _df.dropna(subset=['samples', 'columns'], inplace=True)
+    _df[COL_NAME] = _df.samples * _df['columns']
+    ax = axs[0]
+    # k a b
+    # bmin = _df['MaxRSSRaw'].min()
+    models = [mem_model1, mem_model2, mem_model3, mem_model4]
+    _resource_allocation_plot_helper(_df, ax, cname, sname, "MaxRSSRaw",
+                                     models)
+
+    # bmin = _df['ElapsedRaw'].min()
+    ax = axs[1]
+    models = [time_model1, time_model2, time_model3, time_model4]
+    _resource_allocation_plot_helper(_df, ax, cname, sname, "ElapsedRaw",
+                                     models)
+    return fig, axs
+
+
+def _resource_allocation_plot_helper(_df, ax, cname, sname, curr, models):
+    COL_NAME = 'samples * columns'
+    M1G = 2**30
+    """Helper function for resource allocation plot. Builds plot for MaxRSSRaw
+    and ElapsedRaw
+
+    Parameters
+    ----------
+    _df: pandas dataframe, required
+        Filtered dataframe for the plot
+    ax : matplotlib axes, required
+        Axes for current subplot
+    cname: str, required
+        Specified job type
+    sname: str, optional
+        Specified job sub type.
+    curr: str, required
+        Either MaxRSSRaw or ElapsedRaw
+    models: list, required
+        List of functions that will be used for visualization
+
+    """
+
+    x_data, y_data = _df[COL_NAME], _df[curr]
+    ax.scatter(x_data, y_data, s=2, label="data")
+    d = dict()
+    for index, row in _df.iterrows():
+        x_value = int(row[COL_NAME])
+        y_value = row[curr]
+        if x_value not in d:
+            d[x_value] = []
+        d[x_value].append(y_value)
+
+    for key in d.keys():
+        d[key].sort()
+        d[key] = [d[key][-1] * 1.05]
+
+    x_data = []
+    y_data = []
+
+    # Populate the lists with data from the dictionary
+    for x, ys in d.items():
+        for y in ys:
+            x_data.append(x)
+            y_data.append(y)
+
+    x_data = np.array(x_data)
+    y_data = np.array(y_data)
+
+    # ax.scatter(x_data, y_data, color='purple', s=2)
+
+    ax.set_xscale('log')
+    ax.set_yscale('log')
+    ax.set_ylabel(curr)
+    ax.set_xlabel("samples * columns")
+
+    best_model, options = _resource_allocation_calculate(_df, x_data, y_data,
+                                                         models, curr, 100)
+    k, a, b = options.x
+    x_plot = np.array(sorted(_df[COL_NAME].unique()))
+
+    if curr == "MaxRSSRaw":
+        y_plot = M1G*1.5 + (np.log(x_plot)*M1G*1.5) + (x_plot*3500)
+        ax.plot(x_plot, y_plot, linewidth=1, color='red',
+                label="predefined curve")
+    if curr == "ElapsedRaw":
+        y_plot = 600 + (np.log(x_plot)*3000)
+        ax.plot(x_plot, y_plot, linewidth=1, color='red',
+                label="predefined curve")
```

Comment:
remove the predefined curves, they were nice for the presentation but not needed on the real code